[libc++] Fix `take_view::__sentinel`'s `operator==`

JMazurkiewicz opened the pull request:
* Fix `take_view::__sentinel`'s `operator==`
* Rename `ranges/range.adaptors/range.take/sentinel/base.pass.cpp` directory to `ranges/range.adaptors/range.take/range.take.sentinel/base.pass.cpp`
* Add ***full*** test coverage for `take_view::__sentinel`'s `operator==`
* Drive-by: fix comment in `base.pass.cpp` test
* Close #55211

JMazurkiewicz committed:
[libc++] Fix `take_view::__sentinel`'s `operator==`

cjdb reviewed (approved):
LGTM

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.take/range.take.sentinel/eq.pass.cpp:
```diff
@@ -0,0 +1,192 @@
+static_assert(!std::forward_iterator<StrictIterator<true>>);
+
+template <bool Const>
+class StrictSentinel {
```
I found the name StrictSentinel a bit confusing. It is actually not strict because it supports all the cross-const comparison with iterator.

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.take/range.take.sentinel/eq.pass.cpp:
```diff
@@ -0,0 +1,192 @@
+static_assert(std::ranges::range<NotSimpleView>);
+static_assert(std::ranges::range<const NotSimpleView>);
+
+struct FancyNonSimpleView : std::ranges::view_base {
```
Similarly, I also found the FancyNonSimpleView name a bit misleading. It is actually less fancier than the NonSimpleView. 

IIUC, NonSimpleView is the cross-const comparable view, and
FancyNonSimpleView is the non-cross-const-comparable view

JMazurkiewicz replied:
I've renamed these things to:
* `StrictIterator` -> `MaybeConstIterator`
* `FancyNonSimpleView` -> `NonCrossConstComparableView`
* `NonSimpleView` -> `CrossConstComparableView`
* `StrictSentinel` -> `CrossConstComparableSentinel`

I hope this is ok.

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.take/range.take.sentinel/eq.pass.cpp:
```diff
@@ -0,0 +1,192 @@
+    // Those should be valid
+    static_assert(
+        weakly_equality_comparable_with<std::ranges::iterator_t<TakeView>, std::ranges::sentinel_t<TakeView>>);
+    static_assert(
```
This line looks exactly the same as the line above

JMazurkiewicz replied:
Oops, it was supposed to check `const TakeView`. Fixed.

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.take/range.take.sentinel/base.pass.cpp:
```diff
@@ -8,10 +8,7 @@
-// constexpr explicit sentinel(sentinel_t<Base> end);
-// constexpr sentinel(sentinel<!Const> s)
-//   requires Const && convertible_to<sentinel_t<V>, sentinel_t<Base>>;
+// constexpr sentinel_t<Base> base() const;
```
Not attached to this line.  What is the reason to rename the folder? The folder name is meant to itch the spec section id

JMazurkiewicz replied:
And `range.take.sentinel` is correct name of the spec section: http://eel.is/c++draft/range.take.sentinel.

Zingam replied:
Maybe that's already encoded in the parent directory as all other tests use just "sentinel" and "iterator"

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.take/range.take.sentinel/eq.pass.cpp:
```diff
@@ -0,0 +1,192 @@
+      std::same_as<bool> decltype(auto) b3 = (tv.end() != tv.begin());
+      assert(b3);
+      std::same_as<bool> decltype(auto) b4 = (tv.begin() != tv.end());
+      assert(b4);
```
Nit: Since you are refactoring the test. Could you add another assert where the expression on the right actually return “false” ?

JMazurkiewicz replied:
Added.

huixie90 reviewed (changes requested):
Thank you for the fix! LGTM with some nits

JMazurkiewicz committed:
Check `const TakeView` instead of `TakeView` in (duplicated) static assertion

Comment: https://github.com/llvm/llvm-project/pull/74655#discussion_r1418511110

JMazurkiewicz committed:
Add tests for comparisons returning `false`

Comment: https://github.com/llvm/llvm-project/pull/74655/files#r1418520744

JMazurkiewicz committed:
Turn `StrictIterator` into alias template for `cpp20_input_iterator`

JMazurkiewicz committed:
Rename `StrictIterator` to `MaybeConstIterator`

JMazurkiewicz committed:
Rename `FancyNonSimpleView` to `NonCrossConstComparableView`

Comment: https://github.com/llvm/llvm-project/pull/74655/files#r1418509780

JMazurkiewicz committed:
Rename `NonSimpleView` to `CrossConstComparableView`

JMazurkiewicz committed:
Rename `StrictSentinel` to `CrossConstComparableSentinel`

Comment: https://github.com/llvm/llvm-project/pull/74655#discussion_r1418507136

JMazurkiewicz committed:
Use `"test_comparisons.h"` header

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/take_view_sentinel

JMazurkiewicz committed:
`"test_comparisons.h"`: `constexpr` -> `TEST_CONSTEXPR`

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/take_view_sentinel

huixie90 commented on libcxx/include/__ranges/take_view.h:
```diff
@@ -183,7 +183,7 @@ class take_view<_View>::__sentinel {
-  friend constexpr bool operator==(const _Iter<_Const>& __lhs, const __sentinel& __rhs) {
+  friend constexpr bool operator==(const _Iter<_OtherConst>& __lhs, const __sentinel& __rhs) {
     return __lhs.count() == 0 || __lhs.base() == __rhs.__end_;
   }
```
Nothing to do with your patch, just wanted to point out that we don’t seem to have tests to test the default argument. (you don’t need to do it as this is nothing to do with the bug you are fixing)
template<bool _OtherConst = !_Const>

JMazurkiewicz replied:
I've tried changing default argument to `_Const` and removing it, and in both cases test failed. I think that current coverage is enough.

huixie90 reviewed (approved):
LGTM with green CI

JMazurkiewicz committed:
`"test_comparisons.h"`: `TEST_CONSTEXPR` -> `TEST_CONSTEXPR_CXX14`

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/take_view_sentinel

huixie90 commented:
Hi, could you please apply clang-format to the lines you changed and I am going to merge it after that

JMazurkiewicz committed:
Make `clang-format` happy

Comment: https://github.com/llvm/llvm-project/pull/74655#issuecomment-1853710543

JMazurkiewicz committed:
Merge branch 'main' into libcxx/ranges/take_view_sentinel

JMazurkiewicz commented:
@huixie90 CI is *almost* green (Android failure is probably unrelated).
